src/LanguageServer/Impl/Diagnostics/DiagnosticsService.cs

```diff
@@ -91,7 +110,7 @@ internal sealed class DiagnosticsService : IDiagnosticsService, IDisposable {
 
         private void PublishDiagnostics() {
             lock (_lock) {
-                foreach (var kvp in _diagnostics) {
+                foreach (var kvp in Diagnostics) {
```

Comment:
This loop calls third party method under lock. We can move it outside.

Reply:
Sure. We only need `_changed` there.